formula/src/test/java/com/instacart/formula/FormulaRuntimeTest.kt

```diff
@@ -212,6 +214,113 @@ class FormulaRuntimeTest(val runtime: TestableRuntime, val name: String) {
+                                 */
+                                observer?.input(1)
+                                none()
+                            }
+                        }
+                    }
+                )
+            }
+        }
+
+        observer = runtime.test(root)
+        observer.input(0)
+        observer.output { assertThat(this).isEqualTo(1) }
+
+        // Check that termination was called
+        assertThat(terminationCallback.values()).containsExactly(0).inOrder()
+    }
+
+    @Test fun `runtime termination triggered while formula is running`() {
+        val terminationCallback = TestEventCallback<Int>()
+        var observer: TestFormulaObserver<Int, *, *>? = null
+        val root = object : StatelessFormula<Int, Int>() {
+            override fun Snapshot<Int, Unit>.evaluate(): Evaluation<Int> {
+                return Evaluation(
+                    output = input,
+                    actions = context.actions {
+                        Action.onTerminate().onEvent {
+                            transition {
+                                terminationCallback.invoke(input)
+                            }
+                        }
+
+                        if (input == 0) {
+                            Action.onInit().onEvent {
+                                // This is outside of effect to trigger termination while running
+                                observer?.dispose()
+                                none()
+                            }
+                        }
+                    }
+                )
+            }
+        }
+
+        observer = runtime.test(root)
+        observer.input(0)
+
+        // Check that termination was called
+        assertThat(terminationCallback.values()).containsExactly(0).inOrder()
+    }
+
+    @Test fun `runtime termination triggered by an effect`() {
+        val terminationCallback = TestEventCallback<Int>()
+        var observer: TestFormulaObserver<Int, *, *>? = null
+        val root = object : StatelessFormula<Int, Int>() {
+            override fun Snapshot<Int, Unit>.evaluate(): Evaluation<Int> {
+                return Evaluation(
+                    output = input,
+                    actions = context.actions {
+                        Action.onTerminate().onEvent {
+                            transition {
+                                terminationCallback.invoke(input)
+                            }
+                        }
+
+                        if (input == 0) {
+                            Action.onInit().onEvent {
+                                transition {
+                                    observer?.dispose()
+                                }
+                            }
+                        }
+                    }
+                )
+            }
+        }
+
+        observer = runtime.test(root)
+        observer.input(0)
```

Comment:
Should we assert the output value here? Should be 0, right?

Reply:
There should be no output given formula exited before producing one.

Reply:
Hmm, I thought there would be a single evaluation, with the transition running afterward?

Reply:
There was, but since formula was terminated during transition, we don't emit an output.